gr-soapy/lib/block_impl.cc

```diff
@@ -0,0 +1,651 @@
+            " channels available.");
+    }
+
+    /*
+     * Validate stream arguments for all channels
+     */
+    for (const auto channel : d_channels) {
+        SoapySDR::ArgInfoList supported_args =
+            d_device->getStreamArgsInfo(d_direction, channel);
+        SoapySDR::Kwargs stream_kwargs = SoapySDR::KwargsFromString(stream_args);
+
+        for (const auto& arg : stream_kwargs) {
+            if (!arg_info_has_key(supported_args, arg.first)) {
+                throw std::invalid_argument(name() + ": Unsupported stream argument " +
+                                            arg.first + " for channel " +
+                                            std::to_string(channel));
+            }
+        }
+    }
+
+    /* Validate tuning arguments for each channel */
+    for (const auto channel : d_channels) {
+        SoapySDR::ArgInfoList supported_args =
+            d_device->getFrequencyArgsInfo(d_direction, channel);
+
+        /* If single set of args, apply to all channels */
+        const size_t arg_idx = tune_args.size() == 1 ? 0 : channel;
+        SoapySDR::Kwargs tune_kwargs = SoapySDR::KwargsFromString(tune_args[arg_idx]);
+
+        for (const auto& arg : tune_kwargs) {
+            if (!arg_info_has_key(supported_args, arg.first)) {
+                throw std::invalid_argument(name() + ": Unsupported tune argument " +
+                                            arg.first + " for channel " +
+                                            std::to_string(channel));
+            }
+        }
+        d_tune_args.push_back(tune_kwargs);
+    }
+
+    /* Validate and apply other settings to each channel */
+    for (const auto channel : d_channels) {
+        SoapySDR::ArgInfoList supported_settings =
+            d_device->getSettingInfo(d_direction, channel);
+
+        /* If single set of args, apply to all channels */
+        const size_t arg_idx = other_settings.size() == 1 ? 0 : channel;
+        SoapySDR::Kwargs settings_kwargs =
+            SoapySDR::KwargsFromString(other_settings[arg_idx]);
+
+        /*
+         * Before applying the setting, make a last check if the setting
+         * is a valid gain
+         */
+        std::vector<std::string> gain_names = d_device->listGains(d_direction, channel);
+        for (const auto& gain_name : gain_names) {
+            SoapySDR::Kwargs::iterator iter = settings_kwargs.find(gain_name);
+            if (iter != settings_kwargs.end()) {
+                d_device->setGain(
+                    d_direction, channel, gain_name, std::stod(iter->second));
+                /*
+                 * As only valid settings are applied, we remove it so it does not
+                 * recognized as an invalid setting
+                 */
+                settings_kwargs.erase(iter);
+            }
+        }
+
+        for (const auto& arg : settings_kwargs) {
+            if (arg_info_has_key(supported_settings, arg.first)) {
```

Comment:
Up to you, but I'm a fan of removing needless `else`, and thus less indentation.

```c++
If (!arg_info_has_key…
  throw…
}
d_device->…
```